dotnet/src/InternalUtilities/src/Data/VectorStoreRecordPropertyReader.cs

```diff
@@ -15,73 +14,385 @@ namespace Microsoft.SemanticKernel.Data;
+    public void VerifyDataProperties(HashSet<Type> supportedTypes, HashSet<Type> supportedEnumerableElementTypes)
+    {
+        VectorStoreRecordPropertyVerification.VerifyPropertyTypes(this._dataProperties, supportedTypes, supportedEnumerableElementTypes, "Data");
+    }
+
+    /// <summary>Verify that the types of the vector properties fall within the provided set.</summary>
+    /// <param name="supportedTypes">The list of supported types.</param>
+    public void VerifyVectorProperties(HashSet<Type> supportedTypes)
+    {
+        VectorStoreRecordPropertyVerification.VerifyPropertyTypes(this._vectorProperties, supportedTypes, "Vector");
+    }
+
+    /// <summary>
+    /// Get the storage property name for the given data model property name.
+    /// </summary>
+    /// <param name="dataModelPropertyName">The data model property name for which to get the storage property name.</param>
+    /// <returns>The storage property name.</returns>
+    public string GetStoragePropertyName(string dataModelPropertyName)
+    {
+        return this._storagePropertyNamesMap.Value[dataModelPropertyName];
+    }
+
+    /// <summary>
+    /// Get the name under which a property will be stored if serialized to JSON
+    /// </summary>
+    /// <param name="dataModelPropertyName">The data model property name for which to get the JSON name.</param>
+    /// <returns>The JSON name.</returns>
+    public string GetJsonPropertyName(string dataModelPropertyName)
+    {
+        return this._jsonPropertyNamesMap.Value[dataModelPropertyName];
+    }
 
     /// <summary>
-    /// Verify that the given type has a public parameterless constructor and return it if it exists.
+    /// Check if we have previously loaded the <see cref="PropertyInfo"/> objects from the data model and if not, load them.
     /// </summary>
-    /// <param name="type">The type to get the constructor for.</param>
-    /// <returns>The public parameterless constructor.</returns>
-    /// <exception cref="ArgumentException">Thrown if the provided type doesn't have a public parameterless constructor.</exception>
-    public static ConstructorInfo GetParameterlessConstructor(Type type)
+    private void LoadPropertyInfoIfNeeded()
     {
-        var constructor = type.GetConstructor(Type.EmptyTypes);
-        if (constructor == null)
+        if (this._keyPropertiesInfo != null)
         {
-            throw new ArgumentException($"Type {type.FullName} must have a parameterless constructor.");
+            return;
         }
 
-        return constructor;
+        // If we previously built the definition from the data model, the PropertyInfo objects
+        // from the data model would already be saved. If we didn't though, there could be a mismatch
+        // between what is defined in the definition and what is in the data model. Therefore, this
+        // method will throw if any property in the definition in not on the data model.
```

Comment:
Not sure if it's a typo or not:
```suggestion
        // method will throw if any property in the definition is not in the data model.
```

Reply:
Yes, it's a typo.  Thanks, fixed